### PR TITLE
[WIP] Write-Ahead log and crash recovery

### PR DESCRIPTION
**Pull request by Warchant**
- [x] Write-Ahead log implementation in `RWFileSafe`
- [x] `Env` implementation for POSIX
- [ ] File locks in `RWFileSafe` implementation
- [x] Tests for RWFileSafe
- [ ] Tests for file locks
- [ ] Tests for ENV
- [ ] Tests for recovery mechanism
- [x] Successful build (fails because of `FixedTable`)

### CODE REVIEW

File: src/ametsuchi/file/file.cc [hunk 1]
```diff
@@ -68,93 +70,80 @@ void File::seek_to_end() { std::fseek(file_.get(), 0, SEEK_END); }
...
 
-/////////////////
-/// ReadWriteFile
-ReadWriteFile::ReadWriteFile(const std::string &path) : File::File(path) {
-  read_  = true;
-  write_ = true;
+bool File::clear() {
+  auto deleted = remove();
+  auto opened = open();
+  return deleted && opened;
```
Comment (l4l) [inline src/ametsuchi/file/file.cc]: Just `opened`. File already can be deleted

File: src/ametsuchi/table/field_table.cc [hunk 2]
```diff
@@ -16,56 +16,58 @@
...
 
 #include <ametsuchi/table/field_table.h>
-#include <ametsuchi/table/table.h>
 
 namespace ametsuchi {
 namespace table {
 
-FieldTable::FieldTable(const std::string &p) : f_(p), file_size(0) {f_.open();}
+FieldTable::FieldTable(const std::string &p)
+    : file_size{0}, f_(new file::RWFileSafe(p)) {
```
Comment (l4l) [inline src/ametsuchi/table/field_table.cc]: replace to `file_size(0)` plz

File: test/ametsuchi/table/fixed_table_test.cc [hunk 3]
```diff
@@ -44,11 +41,9 @@ struct ExampleFB {
...
 
 // 15 char s + ending \0
-#define STR_16(c) #c#c#c#c\
-                  #c#c#c#c\
-                  #c#c#c#c\
-                  #c#c#c
-#define FB(c, i, d) ExampleFB{STR_16(c), i, d}
+#define STR_16(c) #c #c #c #c #c #c #c #c #c #c #c #c #c #c #c
+#define FB(c, i, d) \
+  ExampleFB { STR_16(c), i, d }
```
Comment (l4l) [inline test/ametsuchi/table/fixed_table_test.cc]: wtf-alignment
plz don't commit this devil's machinations

File: test/ametsuchi/table/fixed_table_test.cc [hunk 4]
```diff
@@ -64,14 +59,11 @@ TEST_F(FileTest, AppendGetTest) {
 
 TEST_F(FileTest, AppendGetBatchTest) {
   FixedTable<ExampleFB> table(filename);
-  uint64_t i[] = {0x12345678, 0x87654321, 0xfedcba9876543210, 0x123456789abcdef};
+  uint64_t i[] = {0x12345678, 0x87654321, 0xfedcba9876543210,
+                  0x123456789abcdef};
```
Comment (l4l) [inline test/ametsuchi/table/fixed_table_test.cc]: god why

File: test/ametsuchi/table/fixed_table_test.cc [hunk 5]
```diff
@@ -64,14 +59,11 @@ TEST_F(FileTest, AppendGetTest) {
...
   double d[] = {1234567e+89, 9876543e+21, 0xfedcba9876543, 0x123456789abcde};
-  std::vector<ExampleFB> base_fbs = {
-    FB(A, i[0], d[0]),
-    FB(B, i[1], d[1]),
-    FB(C, i[2], d[2]),
-    FB(D, i[3], d[3])
-  };
+  std::vector<ExampleFB> base_fbs = {FB(A, i[0], d[0]), FB(B, i[1], d[1]),
+                                     FB(C, i[2], d[2]), FB(D, i[3], d[3])};
```
Comment (l4l) [inline test/ametsuchi/table/fixed_table_test.cc]: same
that pretty hard to edit w/ multiple cursors

File: src/ametsuchi/table/field_table.cc [hunk 6]
```diff
@@ -77,42 +79,48 @@ ByteArray FieldTable::get(const file::offset_t offset) {
...
+FieldTable::ForwardIterator::ForwardIterator(FieldTable &ft,
+                                             file::offset_t offset)
     : ft_(ft), offset_(offset) {
   value_ = ft_.get(offset);
 }
 
 
-FieldTable::ForwardIterator::ForwardIterator(const FieldTable::ForwardIterator &it)
+FieldTable::ForwardIterator::ForwardIterator(
+    const FieldTable::ForwardIterator &it)
```
Comment (l4l) [inline src/ametsuchi/table/field_table.cc]: this alignment makes me cry

File: include/ametsuchi/file/file.h [hunk 7]
```diff
@@ -55,20 +62,44 @@ class File {
...
 
   bool can_read();
   bool can_write();
 
+  void set_path(const std::string &path);
+
+  std::string get_name();
+
+  /**
+   * Remove file.
```
Comment (Warchant) [inline include/ametsuchi/file/file.h]: This documentation is for doxygen

File: include/ametsuchi/env.h [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    struct sysinfo memInfo;
+    sysinfo(&memInfo);
+    // to avoid overflow on right hand side
+    uint64_t mem = memInfo.totalram;
+    mem += memInfo.totalswap;
+    mem *= memInfo.mem_unit;
+    return mem;
+  }
+
+  std::string mkdirectory(std::string path) {
```
Comment (Warchant) [inline include/ametsuchi/env.h]: Folder and file are not related. We need this only in single place - in env

File: include/ametsuchi/file/file.h [hunk 9]
```diff
@@ -55,20 +62,44 @@ class File {
...
+
+  /**
    * Reads exactly \p size bytes from file at current cursor position.
    * @param size
    * @return byte array of \p size
    */
   ByteArray read(size_t size);
 
+  /**
+   * Get path to file.
```
Comment (Warchant) [inline include/ametsuchi/file/file.h]: This is for doxygen. Every interface method should have description

File: include/ametsuchi/file/rw_file_plain.h [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+#include <ametsuchi/file/rw_file.h>
+#include <string>
+
+namespace ametsuchi {
+namespace file {
+
+/**
+ * RWFilePlain does:
+ *  - read and write files
```
Comment (Warchant) [inline include/ametsuchi/file/rw_file_plain.h]: For you, not for those who will be reading generated doxy files

**Comment by Warchant** [comment]
:rage:

**Comment by Warchant** [comment]
PR is too big. I'll merge it with master to allow others pull changes.
